src/gui/DatabaseWidget.cpp

```diff
@@ -1379,6 +1379,12 @@ bool DatabaseWidget::lock()
         return true;
     }
 
+    // Don't try to lock the database while saving, doing so
+    // would result in data loss and/or crashing
+    if (m_db->isSaving()) {
+        return false;
+    }
```

Comment:
Is this the correct behaviour? To me it would make more sense to defer (or reschedule) further execution to when isSaving() returns false (without blocking the event loop).

Reply:
I tried that and it causes a deadlock because you are in a single thread that cannot unwind itself.

Reply:
This should cause a deadlock if you do not release the lock in isSaving() in any case. However, scheduling reentry using a timer should not result in any deadlocks.

Reply:
I'll draw a diagram of the problem, it is what is causing the majority of crashes.